torchaudio/functional_filtering.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import math
+import torch
+import math
+import torchaudio
+
+"""
+Biquad filtering approach based on SoX implementation 14.4.2
+
+Performance is currently very slow
+"""
+
+__all__ = [
+    'lowpass1',
+    'biquad',
+    'lowpass_biquad',
+    'highpass_biquad',
+]
+
+
+def _dB2Linear(x):
+    return math.exp(x * math.log(10) / 20.0)
+
+def highpass_biquad(input, sr, cutoff_freq, Q = 0.707):
+
+    GAIN = 1
+    w0 = 2 * math.pi * cutoff_freq / sr
+    A = math.exp(GAIN / 40.0 * math.log(10))
+    alpha = math.sin(w0) / 2 / Q
+    mult = _dB2Linear(max(GAIN, 0))
+
+    b0 = (1 + math.cos(w0))/2
+    b1 = -1 - math.cos(w0)
+    b2 = b0
+    a0 = 1 + alpha
+    a1 = -2 * math.cos(w0)
+    a2 = 1 - alpha
+    return biquad(input, b0, b1, b2, a0, a1, a2)
+
+
+def lowpass_biquad(input, sr, cutoff_freq, Q = 0.707):
+    
+    GAIN = 1
+    w0 = 2 * math.pi * cutoff_freq / sr
+    A = math.exp(GAIN / 40.0 * math.log(10))
+    alpha = math.sin(w0) / 2 / Q
+    mult = _dB2Linear(max(GAIN, 0))
+
+    b0 = (1 - math.cos(w0))/2
+    b1 = 1 - math.cos(w0)
+    b2 = b0
+    a0 = 1 + alpha
+    a1 = -2 * math.cos(w0)
+    a2 = 1 - alpha
+    return biquad(input, b0, b1, b2, a0, a1, a2)
+
+
+def biquad(input, b0, b1, b2, a0, a1, a2):
+
+    n_channels, n_samples = input.size()
+    output = torch.zeros_like(input)
+
+    # Normalize the coefficients
+    b2 = b2 / a0
+    b1 = b1 / a0
+    b0 = b0 / a0
+    a2 = a2 / a0
+    a1 = a1 / a0
+
+    # Initial Conditions set to zero
+    pi1 = torch.zeros(n_channels, 1)
+    pi2 = torch.zeros(n_channels, 1)
+    po1 = torch.zeros(n_channels, 1)
+    po2 = torch.zeros(n_channels, 1)
+
+    for i in range(input.size()[1]):
+        o0 = input[:, i] * b0 + pi1 * b1 + pi2 * b2 - po1 * a1 - po2 * a2
+        pi2 = pi1
+        pi1 = input[:, i]
+        po2 = po1
+        po1 = o0
+        output[:, i] = o0
+
+    return output
+
+def lowpass1(waveform,           # type: Tensor
```

Comment:
Let's leave it in for now, but this will be superseded by the other low pass. Out of curiosity, have you seen performance difference between the two low pass filters?

Reply:
I have decided to remove this.  I will do comparisons of performance between FIR and IIR filters.